src/device-manager/java/src/nl/Weave/DataManagement/WDMClientImpl.java

```diff
@@ -0,0 +1,213 @@
+
+import android.os.Build;
+import android.util.Log;
+
+import java.util.EnumSet;
+import java.util.Random;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Iterator;
+
+public class WDMClientImpl implements WDMClient
+{
+    public WDMClientImpl()
+    {
+        init();
+        mCompHandler = null;
+    }
+
+    @Override
+    public void connect(long deviceMgrPtr)
+    {
+        mWDMClientPtr = newWDMClient(deviceMgrPtr);
+    }
+
+    @Override
+    public void close()
+    {
+        if (mTraitMap != null)
+        {
+            Iterator iter = mTraitMap.entrySet().iterator();
+            while (iter.hasNext())
+            {
+                Map.Entry entry = (Map.Entry) iter.next();
+                GenericTraitUpdatableDataSinkImpl traitInstance = (GenericTraitUpdatableDataSinkImpl)entry.getValue();
+                if (traitInstance != null)
+                {
+                    traitInstance.close();
+                }
+            }
+            mTraitMap.clear();
+        }
+
+        if (mWDMClientPtr != 0)
+        {
+            deleteWDMClient(mWDMClientPtr);
+            mWDMClientPtr = 0;
+        }
+
+        mCompHandler = null;
+    }
+
+    @Override
+    public GenericTraitUpdatableDataSinkImpl newDataSink(ResourceIdentifier resourceIdentifier, long profileId, long instanceId, String path)
+    {
+        long traitInstancePtr = 0;
+        boolean isExist = false;
+        GenericTraitUpdatableDataSinkImpl traitInstance = null;
+
+        if (mTraitMap == null)
+        {
+            mTraitMap = new HashMap<Long, GenericTraitUpdatableDataSinkImpl>();
+        }
+
+        if (mWDMClientPtr == 0)
+        {
+            mCompHandler.onError(new Exception("wdmClientPtr is not ready."));
+            return null;
+        }
+
+        traitInstancePtr = newDataSink(mWDMClientPtr, resourceIdentifier, profileId, instanceId, path);
+
+        if (traitInstancePtr == 0)
+        {
+            mCompHandler.onError(new Exception("Fail to create traitInstancePtr."));
+            Log.e(TAG, "traitInstancePtr is not ready");
+            return null;
+        }
+
+        isExist = mTraitMap.containsKey(traitInstancePtr);
+        if (isExist == false)
+        {
+            traitInstance = new GenericTraitUpdatableDataSinkImpl(traitInstancePtr, this);
+            mTraitMap.put(traitInstancePtr, traitInstance);
+        }
+        else
+        {
+            Log.d(TAG, "trait exists" + profileId);
+        }
+
+        Log.d(TAG, "there exists %d traits" + mTraitMap.size());
+
+        return mTraitMap.get(traitInstancePtr);
+    }
+
+    protected void removeDataSinkRef(long traitInstancePtr)
+    {
+        boolean isExist = false;
+        isExist = mTraitMap.containsKey(traitInstancePtr);
+        if (isExist == true)
+        {
+            mTraitMap.put(traitInstancePtr, null);
+        }
+    }
+
+    @Override
+    public void beginFlushUpdate()
+    {
+        if (mWDMClientPtr == 0)
```

Comment:
Same comments here for all of these pointer checks. You can extract a private method that throws an `IllegalStateException` if the pointer isn't initialized and use it throughout.

Reply:
Will do it, thanks

Reply:
Done, thanks